PATTERN.md

```diff
@@ -0,0 +1,289 @@
+be used subsequently as Credentials for GRPC Listening and grpc.ClientConn creation.
+
+### Setup other needed things
+
+If you require a [vpp-agent](https://github.com/ligato/vpp-agent),
+[Example](https://github.com/networkservicemesh/cmd-forwarder-vppagent/blob/9a981c41e568582c3916f8143a826edcd7d8162e/main.go#L84):
+```go
+vppagentCC, vppagentErrCh := vppagent.StartAndDialContext(ctx)
+	exitOnErr(ctx, cancel, vppagentErrCh)
+```
+
+### Instantiate Endpoint
+
+Endpoint instantiation should be simple, generally it will use a chain defined in some sdk*/ repo.
+All configuration parameters should be derived from the config extracted from the environment.
+
+[Example](https://github.com/networkservicemesh/cmd-forwarder-vppagent/blob/9a981c41e568582c3916f8143a826edcd7d8162e/main.go#L100):
+```go
+endpoint := xconnectns.NewServer(
+		config.Name,
+		&authzPolicy,
+		spiffeutils.SpiffeJWTTokenGeneratorFunc(tlsPeer.GetCertificate, config.MaxTokenLifetime),
+		vppagentCC,
+		config.BaseDir,
+		config.TunnelIP,
+		vppinit.Func(config.TunnelIP),
+		&config.ConnectTo,
+		spiffeutils.WithSpiffe(tlsPeer, time.Second),
+		grpc.WithDefaultCallOptions(grpc.WaitForReady(true)),
+	)
+```
+### Create GRPC Server for endpoint
+
+[Example](https://github.com/networkservicemesh/cmd-forwarder-vppagent/blob/9a981c41e568582c3916f8143a826edcd7d8162e/main.go#L114):
+```go
+server := grpc.NewServer(spiffeutils.SpiffeCreds(tlsPeer, time.Second))
+endpoint.Register(server)
+srvErrCh := grpcutils.ListenAndServe(ctx, &config.ListenOn, server)
+```
+
+### Wait for exit 
+
+[Example](https://github.com/networkservicemesh/cmd-forwarder-vppagent/blob/9a981c41e568582c3916f8143a826edcd7d8162e/main.go#L122):
+```go
+<-ctx.Done()
+<-vppagentErrCh
+```
+
+## ./internal
+
+Any pkgs that are local to the repo should be put in ./internal.  If you think those packages should be
+shared with others, then those packages probably belong in an sdk*/ repo.
+
+In the example of cmd-forwarder-vppagent we have three 'internal' packages:
+
+1. ./internal/authz - a helper for authz until a more elegant solution can be provided
+2. ./internal/imports - a simple package used for priming Docker builds to maximize layer caching
+3. ./internal/vppinit - an implementation of the vppinit function that is specific to this cmd
+
+## Dockerfile
+
+Dockerfiles should be kept as simple and highly patterned as possible.  Generally speaking in order to build, test, and debug
+most Network Service Mesh cmds will require:
+
+1. Go - to build
+2. Spire - needed to test
+3. Dlv - neede to debug
+   
+   and then
+4. Build
+5. Test
+6. Debug
+7. Runtime
+
+
+### Go
+If you are working on Network Service Mesh cmds that do not require vppagent, you can usually simply
+use a [golang base docker image](https://hub.docker.com/_/golang) to get Go.  However, in the case of 
```

Comment:
Spire/spiffie is not working with goland base image unfortunately. I've used ubuntu as test image in compare to ligato base image.

Reply:
What’s the issue?

Reply:
@haiodo Have you tried this:
```dockerfile
FROM golang:1.14.3-alpine3.11 as alpine
COPY --from=gcr.io/spiffe-io/spire-agent:0.10.0 /opt/spire/bin /bin
COPY --from=gcr.io/spiffe-io/spire-server:0.10.0 /opt/spire/bin /bin
```
?

The tarball binaries for download are based on glibc.  golang alpine images use musl.  The spiffe-io containers are built on alpine, and thus use musl, and on cursory testing appear to work :)